### ISSUE TITLE
Add --no-glob option for each command accepts glob patterns by default

### ISSUE THREAD

**Issue by EmilyGraceSeville7cf** [nushell/nushell#10364]
### Related problem

Commands like `mv` interpret passed strings as glob patterns out of the box, and there is no way to explicitly disable this behavior AFAIK. Now it's problematic to perform some `mv` operation on a file containing characters can be interpreted as glob like square brackets.

### Describe the solution you'd like

Introduce `--no-glob` option for all commands accept glob patterns by default. Short option name should be resolved case-by-case, but `-n` can be preferred by default.

### Additional context and details

Issue is created out of [this](https://discord.com/channels/601130461678272522/1151559455859802233/1151568201747021844) topic.

**Comment by amtoine** [comment]
this looks like a duplicate of https://github.com/nushell/nushell/issues/9310, right?

will close in favor of https://github.com/nushell/nushell/issues/9310 for now :relieved: 
if it's more than #9310, feel free to reopen :wink:

**Comment by fdncred** [comment]
Seems like this is more of a feature request vs a bug report. If this feature gets implemented, it could close #9310 though.

**Comment by skyvine** [comment]
Hello,

I am new to the nushell codebase but interested in contributing. I started looking at this issue because I originally assumed that it was a blocker for #7579, which has the "help wanted" tag, but after spending some time with it I think that the inverse is actually true: this issue is blocked by #7579 (in particular, standardizing on a single regex library, which is presumably Wax).

I have looked at the code in `ls.rs` as a litmus test to check if this change is something that I have the capacity to manage. In principle I expected it to be a simple change - there is a place where `ls` calls a function to interpret the glob, so we should be able to just put that into a conditional that checks the relevant flag. However, the change is complicated by the logic used to list the contents of a directory. The code currently pushes `*` to the end of the path so that the glob function returns the contents instead of the directory itself. If we skip the glob function completely, then it will look for a file literally named `*`. It would be possible to resolve this by changing the values returned by the `match pattern_arg` from and add some complicated logic around the globbing, but it seems simpler to use `Pattern::escape` on the user's input when they request no globbing.

I have looked at implementing this change in multiple ways. [The current version can be found in my fork](https://github.com/skyvine/nushell-clone). It currently fails the `lists_regular_files_in_special_folder` test. This test exists in the code base, but I added some new cases to it to confirm that my change was correct, and it is one of these cases which failed. The relevant test code is:

```
fn lists_regular_files_in_special_folder() {
    Playground::setup("ls_test_3", |dirs, sandbox| {
        sandbox
            .mkdir("[abcd]")
            .mkdir("[bbcd]")
            .mkdir("abcd]")
            .mkdir("abcd")
            .mkdir("abcd/*")
            .mkdir("abcd/?")
            .with_files(vec![EmptyFile("[abcd]/test.txt")])
            .with_files(vec![EmptyFile("abcd]/test.txt")])
            .with_files(vec![EmptyFile("abcd/*/test.txt")])
            .with_files(vec![EmptyFile("abcd/?/test.txt")])
            .with_files(vec![EmptyFile("abcd/?/test2.txt")]);
...
        let actual = nu!(
            cwd: dirs.test(), format!(r#"ls --glob true [bbcd] | length"#));
        assert!(actual.err.contains("no matches found"));
```

It fails because nu successfully lists the (non-existent) contents of the `[bbcd]` directory. This is the behavior of nu on main, and I believe it is a bug as demonstrated by the following commands. (One of the commands need to be run from a separate shell)

```
(nu) $ mkdir /tmp/sandbox
(nu) $ cd /tmp/sandbox
(nu) $ mkdir a b c d
(nu) $ ls [abcd]
╭───┬──────┬──────┬─────────┬──────────╮
│ # │ name │ type │  size   │ modified │
├───┼──────┼──────┼─────────┼──────────┤
│ 0 │ a    │ dir  │ 4.0 KiB │ now      │
│ 1 │ b    │ dir  │ 4.0 KiB │ now      │
│ 2 │ c    │ dir  │ 4.0 KiB │ now      │
│ 3 │ d    │ dir  │ 4.0 KiB │ now      │
╰───┴──────┴──────┴─────────┴──────────╯
(bash) $ mkdir \[abcd] # in nu, `mkdir [abcd]` and `mkdir [[]abcd[]]` prints the error "expected non-[] value"
(nu) $ ls [abcd]
╭────────────╮
│ empty list │
╰────────────╯
```

As `ls` is supposed to unconditionally interpret `[abcd]` as a glob pattern, the output should not change between the two invocations. The difference occurs because there is some special logic (for getting the contents & checking permissions) which operates on the literal input, regardless of whether or not the user supplied a glob pattern. Conceptually, it is correct to list the directory itself (not the contents) if a glob happens to match exactly 1 file which happens to be a directory. Therefore, this special logic should not run at all if the user input is a glob pattern. Which leads me to the reason why #7579 is a blocker: this change is dependent on the globbing library in use. In the case of `ls`, we need to execute different logic depending on whether or not the input is a glob pattern, regardless of what that pattern expands to. Hopefully all glob libraries have a public helper function to check if some string is a glob pattern making this change simple, but other commands might be dependent in more complex way. Performing this change before standardizing on a single regex engine is likely to result in wasted effort.

On a somewhat unrelated note, the implementation in my branch defines a named parameter `--glob` instead of a `--no-glob` flag. I implemented it this way because I think it would be useful to make the default behavior configurable. At least one user has noted that there are environments where it is common for filenames to contain glob characters (in particular, https://github.com/nushell/nushell/issues/9310#issuecomment-1591899194 notes that next.js uses square brackets in file names). Regardless of the default, the named argument can be used to force a particular interpretation for a particular invocation.

**Comment by fdncred** [comment]
> However, the change is complicated by the logic used to list the contents of a directory. The code currently pushes * to the end of the path so that the glob function ...

This is one of the reasons why I've been looking at `jwalk`, `walkdir`, `ignore`, etc. Another problem is that glob syntax can be used in file naming. So, differentiating a literal `[file].txt` versus a glob of `[file].txt` is kind of hard to do automatically. I'd support having an opt-in solution where the user can escape globbing characters. I'd also support a complete rewrite of `ls` using one of the crates I mention above. I lean toward `jwalk` for it's streaming ability which we cannot lose.

I'd be willing to try `--glob` but I tend to think users would object because the shell is supposed to "just know" what you mean somehow. The globbing issue isn't just related to ls, of course, it's literally every command that interfaces with the file system cp, ucp, mv, rm, touch, etc. It would be nice to have a holistic approach to fixing all this but I'm not sure what that is.

**Comment by skyvine** [comment]
I was looking at the glob-expansion tag and noticed the conversation in #10498. I see that the plan for the future is to use uutils, which presumably includes replacing the current version of `ls` with their version. I am not sure that it would be helpful make any changes to `ls` at this time.

**Comment by fdncred** [comment]
>  ... see that the plan for the future is to use uutils, which presumably includes replacing the current version of ls with their version. I am not sure that it would be helpful make any changes to ls at this time.

I disagree. It remains to be seen whether we'll use coreutils `ls`. It will take a major refactor on their part. So, `ls` is in the `run` category of commands in the `crawl`, `walk`, `run` spreadsheet of coreutils commands. It's labeled `run` because it will be exceedingly difficult (we're guessing). What makes it difficult is `ls` needs to stream data and the coreutils `ls` command does not do that. It needs to somehow iterate through files and folders, yielding output on the way, while still continuing.

I'd love to use their code, I just don't think it will happen any time soon. It's a struggle enough right now to get the `crawl` commands working. Lots of internal changes on the coreutils side just to do that.

So, with that said, I'd rather have an excellent multithreaded `ls` with excellent globbing capabilities, rather than waiting to see if coreutils uu_ls crate ever happens.

**Comment by skyvine** [comment]
> I'd be willing to try --glob but I tend to think users would object because the shell is supposed to "just know" what you mean somehow.

I agree with the point that the shell should "just know" as much as possible. I still think that it is useful to have the `--glob` flag as an escape hatch, so that if a user runs into an unconsidered or ambiguous edge case they can force the behavior that they need.

> So, differentiating a literal [file].txt versus a glob of [file].txt is kind of hard to do automatically.

What would you think about a PR that adds the traditional strategy of "if it doesn't match anything when interpreted as a glob pattern, check if it exists literally" to the `nu_glob` crate?

**Comment by fdncred** [comment]
> "if it doesn't match anything when interpreted as a glob pattern, check if it exists literally"

that sounds like something worth trying to me.

**Comment by skyvine** [comment]
I looked at this approach, and I have a toy implementation of this behavior in [the glob-fallback-to-literal branch of my fork](https://github.com/nushell/nushell/compare/main...skyvine:nushell-clone:glob-fallback-to-literal). I don't intend on developing this into a pull request at this time because the current implementation is not satisfying, and it is not clear to me that a better implementation would be worth the effort before the project settles on a globbing crate. However, it still demonstrates that this approach has some effectiveness as well as some limitations.

## Summary
There are several technical deficiencies with the commit, but most of them could be fixed. The sticking point is that I'm implementing this by just throwing more variables into the code, and adding this clutter will make code difficult to work with over time. Sometimes new variables are needed, but in this case the new member variable impacts every caller in order to support one edge case. The commit adds 2 variables, a new member name `original` to the `Paths` struct and a new local variable named `getting_first_result` in the `next` implementation .

The `original` member is added so that the `next` implementation can access the original input given by the user. It is possible to recover this information by combining the information in the `scope` and `dir_patterns` members, but `scope` is `take`n at the beginning of the first `next` call, so this information is lost. It is not possible to return the literal path in the `take` block because we do not yet know if there are any results. The main part of the `next` implementation is a loop that checks candidates for matches, and returns only once a match is found or it has eliminated all possibilities. Therefore, we cannot know whether there are any matches until we perform at least one iteration through the loop. This is also the reason for the `getting_first_result` flag, because the loop is otherwise unable to determine if there was a previous match.

An alternative strategy I considered was to run the logic at the end of `glob_with` and return a different result if it didn't find any results. However, this would require modifying the return type of `glob_with`. I don't think that we want to `Box` the result, so I could add an `enum` that alternates a `Path` or a `String` and implement `Iterator` on it. Again, this adds a lot of clutter to the file in order to support this one edge case.

There might be a more elegant solution that requires refactoring the code in `nu_glob` in some way. Or perhaps the additional clutter is worth it because this edge case is very important to at least some subset of users. However, I don't think that it makes sense to pursue either of those paths unless and until the project determines that `nu_glob` should be the standard globbing crate moving forward.

I briefly looked at some of the other `nu_glob` callers and didn't see anything that would be obviously problematic about this change, but I did not thoroughly review them due to the above.

## Effectiveness

Most cases "just work" with this change:

```shell
> tree
.
├── [empty]
├── [file]
└── [not-empty]
    ├── [file]
    └── file

3 directories, 3 files
/tmp/sandbox> ls [empty]                                                                                                                                                                                                09/30/2023 10:45:49 AM
╭────────────╮
│ empty list │
╰────────────╯
/tmp/sandbox> ls [file]                                                                                                                                                                                                 09/30/2023 10:45:55 AM
╭───┬────────┬──────┬──────┬─────────────╮
│ # │  name  │ type │ size │  modified   │
├───┼────────┼──────┼──────┼─────────────┤
│ 0 │ [file] │ file │  0 B │ an hour ago │
╰───┴────────┴──────┴──────┴─────────────╯
/tmp/sandbox> ls [not-empty]/file                                                                                                                                                                                       09/30/2023 10:45:59 AM
╭───┬──────────────────┬──────┬──────┬────────────────╮
│ # │       name       │ type │ size │    modified    │
├───┼──────────────────┼──────┼──────┼────────────────┤
│ 0 │ [not-empty]/file │ file │  0 B │ 19 minutes ago │
╰───┴──────────────────┴──────┴──────┴────────────────╯
/tmp/sandbox> ls [not-empty]/[file]                                                                                                                                                                                     09/30/2023 10:46:04 AM
╭───┬────────────────────┬──────┬──────┬────────────────╮
│ # │        name        │ type │ size │    modified    │
├───┼────────────────────┼──────┼──────┼────────────────┤
│ 0 │ [not-empty]/[file] │ file │  0 B │ 19 minutes ago │
╰───┴────────────────────┴──────┴──────┴────────────────╯
```

Globs still work when they match:

```
> tree                                                                                                                                                                                                      09/30/2023 10:50:16 AM
.
├── e
├── s
├── t
└── [test]

1 directory, 4 files
> ls [test]                                                                                                                                                                                                 09/30/2023 10:49:00 AM
╭───┬──────┬──────┬──────┬────────────────╮
│ # │ name │ type │ size │    modified    │
├───┼──────┼──────┼──────┼────────────────┤
│ 0 │ e    │ file │  0 B │ 11 seconds ago │
│ 1 │ s    │ file │  0 B │ 11 seconds ago │
│ 2 │ t    │ file │  0 B │ 11 seconds ago │
╰───┴──────┴──────┴──────┴────────────────╯
```

## Limitations

However, pushing `*` still causes an issue:

```shell
/tmp/sandbox> ls [not-empty]                                                                                                                                                                                            09/30/2023 10:46:27 AM
Error:   × No matches found for [not-empty]/*
   ╭─[entry #10:1:1]
 1 │ ls [not-empty]
   ·    ─────┬─────
   ·         ╰── Pattern, file or folder not found
   ╰────
  help: no matches found
```

Additionally, if there is a literal filename which happens to match other files when interpreted as a glob, there is no way to refer to it directly:

```
> tree                                                                                                                                                                                                      09/30/2023 10:50:16 AM
.
├── e
├── s
├── t
└── [test] # impossible to ls this file directly, must list the whole directory

1 directory, 4 files
> ls [test]                                                                                                                                                                                                 09/30/2023 10:49:00 AM
╭───┬──────┬──────┬──────┬────────────────╮
│ # │ name │ type │ size │    modified    │
├───┼──────┼──────┼──────┼────────────────┤
│ 0 │ e    │ file │  0 B │ 11 seconds ago │
│ 1 │ s    │ file │  0 B │ 11 seconds ago │
│ 2 │ t    │ file │  0 B │ 11 seconds ago │
╰───┴──────┴──────┴──────┴────────────────╯
> ls
╭───┬────────┬──────┬──────┬───────────────╮
│ # │  name  │ type │ size │   modified    │
├───┼────────┼──────┼──────┼───────────────┤
│ 0 │ [test] │ file │  0 B │ 2 minutes ago │
│ 1 │ e      │ file │  0 B │ 2 minutes ago │
│ 2 │ s      │ file │  0 B │ 2 minutes ago │
│ 3 │ t      │ file │  0 B │ 2 minutes ago │
╰───┴────────┴──────┴──────┴───────────────╯
```

**Comment by bobhy** [comment]
I think the original issue is resolved by #10571?

**Comment by horasal** [comment]
I like this `--no-glob` option than any other fixes patching glob's behavier. 

`ls` `cp` `mv` DO accept glob, which means if user gives `[a]` as the input, it should be parsed as glob and program should not guess what user mean. If not there will just be another bug report like `Glob not work when there's a file has the same name with glob pattern`. 

so user should point out how to parse glob-like input, it can be 

* a switch like `--no-glob`
* a special raw string may like `\`string_never_expanded\``, this can be done by expanding current glob type
* a special pattern `[[]` `[]]`, which is already allowed in current version, and it also sholud be pointed out that `\[` is a bad idea because it's broken on windows. Given `c:\myfile\[1]`, It's impossible to know if user means it's a file `myfile[1]` or a glob.

at least any patches may glob inconsistent should be removed from `mv` `ls` `cp` etc.

**Comment by fdncred** [comment]
We talked about this `[]` issue in our team meeting this week and one idea that we're leaning toward right now is to revert the changes I made to nu-glob to bring it inline with glob 3.1. Here's some history.

Way back when, JT forked the glob crate for `ls` and removed the `[]` functionality so the problems with them wouldn't show up. Years later, I merged the latest changes from the glob crate into nu-glob, which they are close to identical now. My thinking was that we'd love to have the `[]` globbing and if there are problems with it, maybe the community could fix them. If we hide the problems, they'd never get fixed.

So, now we're at a crossroad. Should we:
1. revert the changes I made to go back to how it was when JT made the original changes to the glob crate? If we do, it won't fix everything because linux allows `{` and `*` and basically any glob character in file names, so we'll still have problems with those, but it would at least fix the `[]` problems. And we can still use the `glob` command if we want fancy globs.
2. Should we land #10694 
3. Should we not land #10694 and add a `--no-glob`?
4. Should we not land #10694 and add a `raw string literal` syntax?

I'm not really sure which way is right. Maybe we need a poll for this question?

**Comment by ChrisDenton** [comment]
I think one reason this issue is tricky is because it touches on the heart of the nushell philosophy. As I understand it, in the most "pure" form of nu's philosophy the `ls` command would take no arguments (except maybe a string that's interpreted literally). Instead you'd only pipe input and output. So you pipe something that produces a literal path and filter the output using something like `where` (or other filter commands).

The reason for taking arguments is for performance and convenience. Performance because you can maybe use more performant OS operations in some cases and convenience because it's easier if `ls` does more or less the "right" thing by default (e.g. filtering hidden files) without needing to manually add filters. But then this makes `ls` lean slightly more towards a traditional shell utility.

I don't really have a conclusion here. I'm just attempting to articulate why I think figuring out the right solution has been causing a bit of angst.

**Comment by fdncred** [comment]
@ChrisDenton Agreed. One school of thought is that no commands have parameters, and every command is composable via pipelines, but in reality, some commands need parameters.... maybe.

**Comment by skyvine** [comment]
This brings up one aspect of #10694 which I expect to be particularly complicated: how can a user pass configuration to a sub-command within a pipeline?

Some of the flags could be removed with that PR, for example it would make sense to remove `--all` and have a separate element of the pipeline which filters out hidden directories. However, the `--long` flag does not make sense anywhere other than the proposed `fetch-file-metadata` command. So the `ls` function would need to accept this flag *and* know that this flag is supposed to go to `fetch-file-metadata` and not any other command.

In an ideal world, it would be nice to be able to do something like this:

```nu
def ls [pattern $(flags-of fetch-file-metadata) $(flags-of glob)] {
    glob $glob-flags $pattern | fetch-file-metadata $fetch-file-metadata-flags
}
```

With the idea being that nu would inspect the command definitions, dynamically create a matching interface for the function, and store the flags in a variable named by some well-known convention (in the above example, the name of the command appended with `-flags`).  However, I have not looked at the implementation of nu-level functions so I have no idea how much work that would take. Additionally, it's not clear how to handle flags that are present in multiple commands. If the user runs `ls --help`, do they want help for the glob command or the fetch-file-metadata command? More likely, they want help for `ls` itself. We could define some convention so that the user could do something like `ls --glob:help` where the colon separates the namespace from the flag, but this starts seeming like a too-complicated interface.

Another option would just be to not support passing them through. If I want long output for a specific `ls` invocation and that is uncommon for me, I can accept the burden of typing out the whole pipeline every once in a while. If it comes up often enough, I can define a separate function, say `lsl`, which has that flag baked in. This seems more realistic, but it is not clear how well people would receive this decision, and it creates code duplication for the user (if they want to change the glob command used, they now have 2 functions to update). The duplication problem could be mitigated if piplines were inspectable & modifiable objects in `nu` so that the user could do something like:

```nu
def lsl [pattern] {
   unpack-pipline ls | each { if $it.command == ls { create-pipeline-element ls --long $it.arguments } else { $it } } | create-pipeline
}
```

But again, this is a significant scope increase, and that definition is much less readable than a simple pipeline. It would be possible to minimize duplication through helper functions, but this does not scale well because you'd need a different helper for each command that you want to customize.

**Comment by horasal** [comment]
I agree that the "nu" way to pass parameters is through pipe and really like this. Actually it's possible to pass multiple parameters by a record like `{ path: "/home/myname/dir-with-*-in-name", no-glob: true } | ls`. maybe it will be interesting to make commands accept both pipe and positional parameter. 

I think there's no such approach to make everyone happy but it's good to have a design decision pointing out a direction of this problem.

> Maybe we need a poll for this question?

It sounds cool to make everyone involving a design decision :)


-------------------------
(the following is written earlier and I just leave it here FYI) 

From my understand, another feature, typed input, also makes nushell differ from other shell. As glob is already a type in current version (but only act a variant of string now), so IMO it's natural to have a syntax for either creating glob or not parsing into glob.

e.g. 

------------

* A sample syntax for creating glob: `!/"glob_here"/` (just for example, may be some better syntax) and any other input is parsed as string and does not expanded

Pros: 

* Glob function (wax or nu_glob) now only binds to glob syntax and can be expanded anywhere when user give a glob type. Consistent and modular
* convenience when user does not use glob too much

Cons:

* can't just `ls **/*` but should write `ls !/"**/*"/` which is ugly
* It's not compatible with bash (many issues around glob arise because it does not work like bash) and need documentation

--------

* A sample syntax for creating raw-string: `` `no[t] expanded` ``. As `cp` `ls` `mv` accept both glob type and string type, it's possible to only let glob expanded and string type is processed as-is, and in parsing phase, make raw-string `` ` ` `` does not parse into glob.

Pros:

* easy to understand and user may try it even without searching document
* convenience when user user glob more than raw string

Cons

* not so modular as glob function still exists in every command accept glob pattern
* It's still not compatible with bash

**Comment by bobhy** [comment]
I'm very confused by the idea raised here that "pure" nushell shouldn't need arguments.  Is there some background about this on Discord or somewhere I could catch up with before offering more opinion?

Re this PR, I get it: there competing philosophies prompting competing proposals: an explicit `--no-glob` switch vs the implicit behavior of this PR.  Although "explicit is better than implicit", convenience counts for something too and I'm guided by "Nushell: a new type of shell" tagline in opting for convenience.  I'm aware convenience is in the eye of the beholder.

So let's take a scenario and evaluate it under both approaches.

First, assume the user has existing files where one is a path  with glob metachars which would parse as a glob pattern and the others have paths that would match the glob.  The original discord thread raised this problem with files with matching `[]`, but here's a simple case using `*`.  Assume user has files `a*b, axb` and `ayb`. (Though `*` is invalid in Windows paths).  

User types `cp a*b dest/`, not aware of the potential ambiguity in the directory.  

Under `--no-glob`:
1. user issues command `cp a*b dest/`, thinking this will copy `axb` and `ayb`
2. user (eventually) discovers that only file `a*b` was copied.  What does s/he think about this behavior?  
It's unexpected, surely.  User is unhappy that the 2 expected files were not copied.  But user can see that there really was a file `a*b` that got copied. User still wants to copy those other files.
3. User refers to help (having ambiguous paths in the directory is a rare situation) and types `cp --noglob a*b dest/` Q.E.D.

Under the current PR:
1. user issues command `cp a*b dest/`, thinking this will copy `axb` and `ayb`
2. user (eventually) discovers that only file `a*b` was copied.  What does s/he think about this behavior?  
It's unexpected, surely.  User is unhappy that the 2 expected files were not copied.  But user can see that there really was a file `a*b` that got copied.   User expects the `cp` command to copy files, not thinking that arguments to `cp` command must be globs.  
3. So user grumbles a bit then either:
* removes the file `a*b` ("where did *that* come from?"), and repeats `cp a*b dest/` and it works as expected, or,
* grumbles a bit more, thinks for a bit, and types: `cp a?b dest/`, which (in this case) works as expected.  I assert there is *always* a glob that can be constructed to match the desired files and exclude the glob-pattern-named file, or to copy it along with the other files.  

I will agree that step 3 is harder on the user under the current PR.  If s/he really wants to retain the glob-patterned-file *and* copy only the files that should match the glob, s/he has to think a bit harder.  But that's only under a deliberately extreme scenario. 
 In most cases, files don't have any metachars.  When they do, next most likely is that they either don't form valid glob patterns or they don't shadow actual other files that would match that pattern.  In those next most likely cases, the copy command just works as expected and the user didn't have to keep globbing edge cases front of mind.  So I assert that makes this design a pragmatic choice suitable for its intended audience.

Note, the `no-glob` proposal hasn't been coded yet and may have unforeseen problems.  For example, how would you pass 2 arguments to the command, one a literal glob argument but the other a pattern to be expanded?  You can make `--no-glob` a simple switch which applies to all arguments, but if you make it a named value argument  (`--no-glob='a*b'`) it can only be specified once.  Meaning that the user might end up having to issue multiple commands because of a restriction in the design.

Why is this being so hard to get right?

We're all familiar with globbing in the context of unix shells which expand patterns before invoking a command and regardless of which command they invoke.  But nushell (like powershell and CMD.exe) always passes the pattern into the command.  So we should be cautious when applying our mental model of "glob" to nushell.  
When using a globbing shell, we expect to quote the pattern if we want instead to access a single file that has glob metacharacters.  But quoting is much harder on Windows, you can't use `\` and must resort to the arcane `[]`.  Yet Nushell wants to be windows-friendly, so it needs to invent something new that can work the same way in all environments.

**Comment by ChrisDenton** [comment]
IMHO, the nushell philosophy is very similar to the UNIX philosophy but taken to its logical conclusion. The central point is that commands should "do one thing and do it well". If you want something that does more than one thing then commands should "work together" (i.e. be composable, which in nushell usually means using pipelines).

If we take the "one thing" part literal then we can say that using `cp a*b dest/` is currently three (or more) commands. In order to expand the glob it first needs to list the directory and then filter the results based on the pattern. Only then can it do its "one thing" of copying files.

So replacing `cp a*b dest/` with commands that do one thing might look something like (**note**: this is a hypothetical and will not work today):

```
ls | get name | where pattern 'a*b' | cp dest/
```

- `ls` lists the directory.
- `get name` converts the table into a list of file names (since we're not using any other data here). We could alternatively use `$in.name`.
- `where pattern 'a*b'` is something I invented as a shorthand for filtering the list based on a glob-like pattern (`a*b` in this case).
- `cp dest/` takes a path or list of paths as input and copies them to `dest/`. Note that we could say `cp` just takes a single path and then the pipeline would need to more explicitly run `cp` for each input.

The advantage of this is that the `cp` command doesn't have to know anything about pattern matching or listing directories. Also it's clear if you're intending to use a literal file name or matching a pattern; there's no ambiguity. The disadvantage is it's more verbose. Though if the `glob` command is simply an alias for a pipeline that calls `ls` and filters on a pattern then we can stick with the "one thing" philosophy and still simplify things a bit:

```
glob a*b | cp dest/
```

Final note: I'm not trying to advocate for any particular syntax here. I'm just trying to illustrate a possible approach in general terms.

**Comment by fdncred** [comment]
Just writing this down so I don't forget. 

Chris's example makes me think that we already have `where pattern` which is `where name =~ blah`. In this case `=~` means match a regex pattern and `!~` means don't match regex pattern. What if we had a glob operator so you could do `where name =% blah` with `=%` and `!%` representing match glob and don't match glob? I wonder if that would be weird or helpful?

We also have `str expand` that kind of plays in this same arena although it doesn't care if what you give it is a filename or not. It works with strings.

I tend to agree that the 'do one thing well' mantra fits very well with nushell, but on the flip side, people would freak because muscle memory is to do `ls glob`, `cp glob`, `mv glob`, `rm glob`, etc.

Right now, my personal opinion is to land bob's PR and see how it goes, but I'm not sure how the other maintainers feel. We won't do that, if we do that, until after the release on Oct-17.

**Comment by ChrisDenton** [comment]
I'd also add it needn't necessarily be an either/or situation. Like `cp` could be an alias that wraps a `file copy` command with a pipeline that expands globs. So you have the "muscle memory" shortcut for doing the quick thing but for everything else there's the new way.

Of course having multiple ways of doing things has its own downsides. But then having it be just an alias mitigates this a bit by allowing people to see what the shortcut expands to and allowing them to customize it to suit their preferences.

**Comment by WindSoilder** [comment]
Currently I'm trying to use quotes to disable glob.

Once the #11621 is ok, you can quote argument to disable glob.  Examples:
```shell
ls a[bc]d   # apply glob

mkdir ii
cp "a[bc]d" e[fg]h ii  # don't apply glob for `a[bc]d`, apply glob for `e[fg]h`
```

**Comment by bobhy** [comment]
I'm sorry, but i think this would be a significant breaking change for
Windows users, where paths with spaces are the norm.  eg, user *does* want
to glob this command:

```
cp "My Documents/*" backup/
```

**Comment by bobhy** [comment]
The "..." spelliing to quote spaces is normative in Windows CMD and
Powershell, fwiw.

There's already a  quoting scheme for globs, too.  Just enclose a metachar
in `[]`, basically making it into a character classe that matches only one
character.

I think it's working correctly:

```
> touch a*b a?b a^b  # touch command doesn't glob
--------------------------------------------------- ~/work/globtest
------------------------------------------------ |V|> ls
╭───┬──────┬──────┬──────┬──────────╮
│ # │ name │ type │ size │ modified │
├───┼──────┼──────┼──────┼──────────┤
│ 0 │ a*b  │ file │  0 B │ now      │
│ 1 │ a?b  │ file │  0 B │ now      │
│ 2 │ a^b  │ file │  0 B │ now      │
╰───┴──────┴──────┴──────┴──────────╯
--------------------------------------------------- ~/work/globtest
------------------------------------------------ |V|> cp a[*]b star  #
matches a single file.
--------------------------------------------------- ~/work/globtest
------------------------------------------------ |V|> cp a[?]b ques
--------------------------------------------------- ~/work/globtest
------------------------------------------------ |V|> cp a[^]b upca
--------------------------------------------------- ~/work/globtest
------------------------------------------------ |V|> ls
╭───┬──────┬──────┬──────┬────────────────╮
│ # │ name │ type │ size │    modified    │
├───┼──────┼──────┼──────┼────────────────┤
│ 0 │ a*b  │ file │  0 B │ a minute ago   │
│ 1 │ a?b  │ file │  0 B │ a minute ago   │
│ 2 │ a^b  │ file │  0 B │ a minute ago   │
│ 3 │ ques │ file │  0 B │ 18 seconds ago │
│ 4 │ star │ file │  0 B │ 37 seconds ago │
│ 5 │ upca │ file │  0 B │ now            │
╰───┴──────┴──────┴──────┴────────────────╯
--------------------------------------------------- ~/work/globtest
------------------------------------------------ |V|> rm a[?]b
--------------------------------------------------- ~/work/globtest
------------------------------------------------ |V|> ls  # removed only a
single file
╭───┬──────┬──────┬──────┬────────────────╮
│ # │ name │ type │ size │    modified    │
├───┼──────┼──────┼──────┼────────────────┤
│ 0 │ a*b  │ file │  0 B │ a minute ago   │
│ 1 │ a^b  │ file │  0 B │ a minute ago   │
│ 2 │ ques │ file │  0 B │ 36 seconds ago │
│ 3 │ star │ file │  0 B │ a minute ago   │
│ 4 │ upca │ file │  0 B │ 20 seconds ago │
╰───┴──────┴──────┴──────┴────────────────╯
--------------------------------------------------- ~/work/globtest
------------------------------------------------ |V|> ls a[*]*  # ls also
uses glob escaping
╭───┬──────┬──────┬──────┬───────────────╮
│ # │ name │ type │ size │   modified    │
├───┼──────┼──────┼──────┼───────────────┤
│ 0 │ a*b  │ file │  0 B │ 2 minutes ago │
╰───┴──────┴──────┴──────┴───────────────╯
--------------------------------------------------- ~/work/globtest
------------------------------------------------ |V|> ls a[*]
Error:   × No matches found for a[*]
   ╭─[entry #89:1:1]
 1 │ ls a[*]
   ·    ──┬─
   ·      ╰── Pattern, file or folder not found
   ╰────
  help: no matches found

```

Do we really need an alternative scheme?

**Comment by fdncred** [comment]
> The "..." spelliing to quote spaces is normative in Windows CMD and Powershell, fwiw.

That's fine but nushell doesn't work this way unfortunately. only backtick quotes help with spaces in paths and filenames it seems.

**Comment by WindSoilder** [comment]
Hi, @bobhy I'm sorry, I've update comment about the case to use backtick quote.

And would you please to make code part of comment more readable?  I think something is not right here.
![image](https://github.com/nushell/nushell/assets/22256154/ee578b99-65c6-417d-a358-c193c9848580)

I think I'm in fans of `bash/fish` shell way, we need to have a way to glob or not glob inputs.

**Comment by bobhy** [comment]
[reposted above comment: note to self: do not reply to github with code sections!]

The "..." spelliing to quote spaces is normative in Windows CMD and
Powershell, fwiw.

There's already ~~a quoting~~__an escaping__ scheme for globs, too.  Just enclose a metachar
in `[]`, basically making it into a character class that matches only one
character.

I think it's working correctly:

```
> touch a*b a?b a^b  # touch doesn't glob its args
> ls
╭───┬──────┬──────┬──────┬──────────╮
│ # │ name │ type │ size │ modified │
├───┼──────┼──────┼──────┼──────────┤
│ 0 │ a*b  │ file │  0 B │ now      │
│ 1 │ a?b  │ file │  0 B │ now      │
│ 2 │ a^b  │ file │  0 B │ now      │
╰───┴──────┴──────┴──────┴──────────╯
> cp a[*]b star  # single char class escapes the metachar
> cp a[?]b ques
> cp a[^]b upca
> ls
╭───┬──────┬──────┬──────┬────────────────╮
│ # │ name │ type │ size │    modified    │
├───┼──────┼──────┼──────┼────────────────┤
│ 0 │ a*b  │ file │  0 B │ a minute ago   │
│ 1 │ a?b  │ file │  0 B │ a minute ago   │
│ 2 │ a^b  │ file │  0 B │ a minute ago   │
│ 3 │ ques │ file │  0 B │ 18 seconds ago │
│ 4 │ star │ file │  0 B │ 37 seconds ago │
│ 5 │ upca │ file │  0 B │ now            │
╰───┴──────┴──────┴──────┴────────────────╯
> rm a[?]b  # matched only a single file
> ls
╭───┬──────┬──────┬──────┬────────────────╮
│ # │ name │ type │ size │    modified    │
├───┼──────┼──────┼──────┼────────────────┤
│ 0 │ a*b  │ file │  0 B │ a minute ago   │
│ 1 │ a^b  │ file │  0 B │ a minute ago   │
│ 2 │ ques │ file │  0 B │ 36 seconds ago │
│ 3 │ star │ file │  0 B │ a minute ago   │
│ 4 │ upca │ file │  0 B │ 20 seconds ago │
╰───┴──────┴──────┴──────┴────────────────╯
> ls a[*]*  # ls also obeys this escaping rule
╭───┬──────┬──────┬──────┬───────────────╮
│ # │ name │ type │ size │   modified    │
├───┼──────┼──────┼──────┼───────────────┤
│ 0 │ a*b  │ file │  0 B │ 2 minutes ago │
╰───┴──────┴──────┴──────┴───────────────╯
> ls a[*]
Error:   × No matches found for a[*]
   ╭─[entry #89:1:1]
 1 │ ls a[*]
   ·    ──┬─
   ·      ╰── Pattern, file or folder not found
   ╰────
  help: no matches found
```

Do we really need an alternative scheme?

**Comment by WindSoilder** [comment]
> There's already a quoting__an escaping__ scheme for globs, too. Just enclose a metacha in [], basically making it into a character class that matches only one character.

I have to say it feels strange to me...To me, follow same rules to other shells make it easier to use in this case.
